airflow/contrib/operators/kubernetes_operator.py

```diff
@@ -0,0 +1,93 @@
+    :type kubernetes_job_yaml_dictionary: dictionary
+    :kubernetes_job_yaml_template: string representation of a Job yaml,
+        if None will use default
+    :type kubernetes_job_yaml_template: string
+    """
+    def __init__(self,
+                 kubernetes_job_name,
+                 kubernetes_job_yaml_dictionary=None,
+                 kubernetes_job_yaml_template=None,
+                 *args,
+                 **kwargs):
+        super(KubernetesJobOperator, self).__init__(*args, **kwargs)
+        self.kubernetes_job_name = kubernetes_job_name
+        self.kubernetes_job_yaml_dictionary = kubernetes_job_yaml_dictionary or {}
+        self.kubernetes_job_yaml_template = kubernetes_job_yaml_template
+
+    def clean_up(self, error_message=None):
+        """
+        Deleting the job removes the job and all related pods.
+        Iff error_message, raise Exception.
+        """
+        result = subprocess.check_output(args=['kubectl', 'delete', 'job', self.unique_job_name])
+        logging.info(result)
+        if error_message:
+            raise Exception(error_message)
+
+    def on_kill(self):
+        """
+        Run clean up.
+        Raising an error will fail the KubernetesJobOperator task.
+        """
+        self.clean_up(error_message='Job %s was killed.' % self.unique_job_name)
+
+    def poll_job_completion(self):
+        """
+        Polls for Job completion every 60 seconds.
+        Any Failed pods will raise an error and fail the KubernetesJobOperator task.
+        """
+        logging.info('Polling for completion of job: %s' % self.unique_job_name)
+        running_job_count = 1
+        while running_job_count > 0:
+            job_description = subprocess.check_output(args=['kubectl', 'describe', 'job', self.unique_job_name])
+            matched = re.search(r'(\d+) Running / \d+ Succeeded / (\d+) Failed', job_description)
+            logging.info('Current status is: %s' % matched.group(0))
+
+            running_job_count = int(matched.group(1))
+            failed_job_count = int(matched.group(2))
+            # If any Jobs fail, fail KubernetesJobOperator task
+            if failed_job_count != 0:
+                return '%s has failed pods, failing task.' % self.unique_job_name
+
+            time.sleep(SLEEP_TIME_BETWEEN_POLLING)
+
+    def execute(self, context):
+        self.unique_job_name = '%s-%s' % (self.kubernetes_job_name, uuid.uuid4())
+        self.kubernetes_job_yaml_dictionary['job_name'] = self.unique_job_name
+
+        yaml = generate_yaml(self.kubernetes_job_yaml_dictionary,
+                             self.kubernetes_job_yaml_template)
+
+        with tempfile.NamedTemporaryFile(suffix='.yaml') as f:
+            f.write(yaml)
+            f.flush()
+            result = subprocess.check_output(args=['kubectl', 'apply', '-f', f.name])
+            logging.info(result)
+
+        message = self.poll_job_completion()
+        self.clean_up(error_message=message)
```

Comment:
I think this should probably just call clean_up, then `if message: raise Exception(message)`

An alternative: If `poll_job_completion` raised an exception instead, you should wrap this in `try/finally`:

```python
try:
    self.poll_job_completion()
finally:
    self.clean_up()
```

This would have an advantage that no matter what kind of exception is raised by `poll_job_completion`, it will still try to clean it up. For example, `subprocess.check_output` could raise an error. In your current version: clean_up would not be called.

I think this might actually be better, now that I think about it.

Reply:
Great point, fixing this.